lib/Driver/ToolChains/CommonArgs.cpp

```diff
@@ -870,6 +870,21 @@ tools::ParsePICArgs(const ToolChain &ToolChain, const ArgList &Args) {
     PIP = true;
   }
 
+  // Pagerando and '-ffunction-sections'/'-fdata-sections' are incompatible
+  bool UseSeparateSections = isUseSeparateSections(Triple);
+  bool FunctionSections = Args.hasFlag(options::OPT_ffunction_sections,
+                                       options::OPT_fno_function_sections,
+                                       UseSeparateSections);
+  bool DataSection = Args.hasFlag(options::OPT_fdata_sections,
+                                  options::OPT_fno_data_sections,
+                                  UseSeparateSections);
+  if (PIP && (FunctionSections || DataSection))
+    ToolChain.getDriver().Diag(
+        diag::err_drv_pagerando_incompatible_with_separate_sections);
+  // Pagerando requires '-flto'
+  if (PIP && !ToolChain.getDriver().isUsingLTO())
```

Comment:
Pagerando is not (yet) compatible with ThinLTO. We need to check that `getLTOMode() == LTOK_Full`.